### ISSUE TITLE
beforeInteractive scripts duplicated during SSR

### ISSUE THREAD

**Issue by felthy** [vercel/next.js#53651]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 22.6.0: Wed Jul  5 22:21:56 PDT 2023; root:xnu-8796.141.3~6/RELEASE_X86_64
    Binaries:
      Node: 16.15.0
      npm: 8.5.5
      Yarn: 1.22.19
      pnpm: 8.5.1
    Relevant packages:
      next: 12.3.4
      eslint-config-next: 12.3.4
      react: 17.0.2
      react-dom: 17.0.2
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

Script optimization (next/script)

### Link to the code that reproduces this issue or a replay of the bug

https://codesandbox.io/p/sandbox/next-js-duplicated-beforeinteractive-scripts-x2r8k3?file=%2Fapp%2Fpage.tsx%3A1%2C1

### To Reproduce

Render the same `<Script strategy="beforeInteractive">` more than once within a single page. The script will be inserted that many times into the page, so will execute that many times on the browser.

### Describe the Bug

During SSR, `<Script strategy="beforeInteractive" />` renders a `<script>` into the page without de-duping, so the same script executes multiple times on the client. This may occur if a component that renders the `<Script>` is rendered more than once in the component tree.

To fix, the `Script` component should check whether the given `id`/`src` is already present in the `scripts[strategy]` array before it concatenates a new entry and calls `updateScripts(scripts)`.

### Expected Behavior

Only one `<script>` should be rendered to the page during SSR, no matter how many times a `<Script>` is rendered with the same `id` or `src`

**Comment by peterwiebe** [comment]
I am experiencing this issue as well, have not been able to determine the root cause. @felthy do you have a workaround?

**Comment by felthy** [comment]
@peterwiebe yes I have a workaround but it's only for Next.js v12. I was starting a PR but realised the v13 solution will be different and I don't currently have a use case for it (so I wouldn't be able to test it properly without investing a fair bit of extra time).

```typescript
import type { ScriptProps } from 'next/script';

// Workaround for https://github.com/vercel/next.js/issues/53651
function SafeScript({ id, src, strategy, ...restProps }: ScriptProps) {
  // Context is available only during SSR
  const context = useContext(HeadManagerContext);
  if (strategy === 'beforeInteractive' || strategy === 'worker') {
    if (context.updateScripts) {
      const scripts = context.scripts as Record<string, ScriptProps[]>;
      const existing = scripts[strategy] || [];
      if (existing.find((state) => (id ? state.id === id : state.src === src))) {
        return null;
      }
    }
  }
  return <Script id={id} src={src} strategy={strategy} {...restProps} />;
}
```

And use `<SafeScript>` as a drop-in replacement for `<Script>` when you need it.
